scripts/gas-refund-program/transactions-indexing/successful-swap-tx-indexing.ts

```diff
@@ -182,36 +201,83 @@ export async function computeSuccessfulSwapsTxFeesRefund({
       );
 
-      const currGasFeePSP = currGasUsedChainCur.dividedBy(pspRateSameDay);
+      const currGasUsedUSD = currGasUsedChainCur
+        .multipliedBy(currencyRate.chainPrice)
+        .dividedBy(10 ** 18); // chaincurrency always encoded in 18decimals
+
+      const accumulatedGasUsedUSD = currGasUsedUSD.plus(
+        swapperAcc?.accumulatedGasUsedUSD || 0,
+      );
+      const currGasFeePSP = currGasUsedChainCur.dividedBy(
+        currencyRate.pspToChainCurRate,
+      );
 
       const accumulatedGasUsedPSP = currGasFeePSP.plus(
         swapperAcc?.accumulatedGasUsedPSP || 0,
       );
 
       const totalStakeAmountPSP = swapperStake.toFixed(0); // @todo irrelevant?
       const refundPercent = getRefundPercent(totalStakeAmountPSP);
+
       assert(
         refundPercent,
         `Logic Error: failed to find refund percent for ${address}`,
       );
-      const currRefundedAmountPSP = currGasFeePSP.multipliedBy(refundPercent);
+
+      let currRefundedAmountPSP = currGasFeePSP.multipliedBy(refundPercent);
+
+      let currRefundedAmountUSD = currRefundedAmountPSP
+        .multipliedBy(currencyRate.pspPrice)
+        .dividedBy(10 ** 18); // psp decimals always encoded in 18decimals
+
+      if (
+        GRPSystemGuardian.totalRefundedAmountUSD(address)
+          .plus(currRefundedAmountUSD)
+          .isGreaterThan(MAX_USD_ADDRESS_BUDGET)
+      ) {
+        currRefundedAmountUSD = MAX_USD_ADDRESS_BUDGET.minus(
+          GRPSystemGuardian.totalRefundedAmountUSD(address),
+        );
+
+        assert(
+          currRefundedAmountUSD.isGreaterThanOrEqualTo(0),
+          'Logic Error: quantity cannot be negative, this would mean we priorly refunded more than max',
+        );
+
+        currRefundedAmountPSP = currRefundedAmountUSD
+          .dividedBy(currencyRate.pspPrice)
+          .multipliedBy(10 ** 18);
+      }
+
+      GRPSystemGuardian.increaseTotalAmountRefundedUSDForAccount(
+        address,
+        currRefundedAmountUSD,
+      );
+
+      GRPSystemGuardian.increaseTotalPSPRefunded(currRefundedAmountPSP);
 
       const accRefundedAmountPSP = currRefundedAmountPSP.plus(
         swapperAcc?.refundedAmountPSP || 0,
       );
 
+      const refundedAmountUSD = currRefundedAmountUSD.plus(
+        swapperAcc?.refundedAmountUSD || 0,
+      );
+
       const pendingGasRefundDatum: PendingEpochGasRefundData = {
         epoch,
         address,
         chainId,
         accumulatedGasUsedPSP: accumulatedGasUsedPSP.toFixed(0),
         accumulatedGasUsed: accumulatedGasUsed.toFixed(0),
+        accumulatedGasUsedUSD: accumulatedGasUsedUSD.toFixed(0),
         accumulatedGasUsedChainCurrency:
           accumulatedGasUsedChainCurrency.toFixed(0),
         firstBlock: swapperAcc?.lastBlock || swap.blockNumber,
         lastBlock: swap.blockNumber,
         totalStakeAmountPSP,
         refundedAmountPSP: accRefundedAmountPSP.toFixed(0),
+        refundedAmountUSD: refundedAmountUSD.toFixed(),
```

Comment:
It doesn't matter, but let's stick to `toFixed(0)` like you do in other places

Reply:
Laser precise refund ^^'

Reply:
Dollar values are in decimals and it can be fairly low (<1) so we do toFixed(0) will pushing 0 in the DB which is kinda weird.